SELECT FOR JSON Move Enum declaration to extensions

anujsarda3 opened the pull request:
### Description

- Moving enum declaration for FOR JSON from engine's parser.h to extensions parser.h
- This is not a complete fix for the issue

Extensions PR - https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/833

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>
 
### Issues Resolved

BABEL-3691

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

anujsarda3 committed:
SELECT FOR JSON support for various datatypes

- Added support for bit dataype
- Created a function to support the date time datatype format for select
  for json apth

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

Sairakan commented on src/backend/utils/adt/json.c:
```diff
@@ -338,12 +354,67 @@ datum_to_json(Datum val, bool is_null, StringInfo result,
 			break;
 		default:
 			outputstr = OidOutputFunctionCall(outfuncoid, val);
+			// The outfuncoid are specific to babelfish and therefore no macro is defined.
```
please state which data types these are referring to - sys.DATETIME, etc

anujsarda3 replied:
Resolving all the comments as just moving declaration from parser.h to gram-tsql-prologue.y.h

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -150,14 +152,28 @@ json_categorize_type(Oid typoid,
 	/* Look through any domain */
 	typoid = getBaseType(typoid);
 
+	if (sql_dialect==SQL_DIALECT_TSQL)
```
In here, the result should depend on the type, not sql_dialect. So, please add this logic to the below switch-case. In other words, even in postgres dialect, if typoid is `sys.bit` then it should behave the same as in T-SQL.

anujsarda3 replied:
Resolving all the comments as just moving declaration from parser.h to gram-tsql-prologue.y.h

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -150,14 +152,28 @@ json_categorize_type(Oid typoid,
+			 * These OID are specifically for babelfish and therefore no macro is defined.
+			 * This switch case is to handle Bit dataype, to convert the typoid to BOOLOID.
+			 */
+			case 17056:
```
We cannot use a hard-coded oid because it is not a fixed number.

Sairakan replied:
@2jungkook do you know of a way to get the Babelfish OID from within the engine code? I was wondering about this as well

2jungkook replied:
You may need to introduce a new hook, e.g., `lookup_tsql_datatype_oid_hook` (Ref: typecode.c).

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -150,14 +152,28 @@ json_categorize_type(Oid typoid,
 	 * types where there's a cast to json.
 	 */
-
+	
```
Please avoid unnecessary changes.

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -338,12 +354,67 @@ datum_to_json(Datum val, bool is_null, StringInfo result,
 		default:
 			outputstr = OidOutputFunctionCall(outfuncoid, val);
+			// The outfuncoid are specific to babelfish and therefore no macro is defined.
+			if ((sql_dialect==SQL_DIALECT_TSQL) && 
```
1. We should not depend on sql_dialect.
2. Our policy is to minimize PG Engine's change. Let's introduce a hook here.

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -338,12 +354,67 @@ datum_to_json(Datum val, bool is_null, StringInfo result,
+			{
+				StringInfo format_output = makeStringInfo();
+				for_json_path_date_time_format(format_output,outputstr, outfuncoid);
+				outputstr = format_output->data;
```
1. `outputstr` returned from line#356 could be leaked.
2. `format_output->data` could be double-freed because of line#369 and freeing from `StringInfo`.

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -338,12 +354,67 @@ datum_to_json(Datum val, bool is_null, StringInfo result,
+ * This funciton handles the format for date & time datatypes by converting the output
+ * into required format for SELECT FOR JSON PATH. For example:
+ * "2022-11-11 20:56:22.41" -> "2022-11-11T20:56:22.41" for datetime, datetime2 & smalldatetime
+ * "2022-11-11 22:25:01.015 +00:00" -> "2022-11-11T22:25:01.015Z" for datetimeoffset
```
Please add more tests, e.g., `2022-11-11 22:25:01.015 +01:00`

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -338,12 +354,67 @@ datum_to_json(Datum val, bool is_null, StringInfo result,
+ * "2022-11-11 22:25:01.015 +00:00" -> "2022-11-11T22:25:01.015Z" for datetimeoffset
+ */
+static void
+for_json_path_date_time_format(StringInfo format_output, char *outputstr, Oid outfuncoid)
```
I think `for_json_datetime_format` is a more proper name because
1. It can be used for both `for_json_path` and `for_json_auto`
2. This is about `datetime`-related data type, not `date` and `time`. Note that `date` and `time` are different data types.

2jungkook commented on src/backend/utils/adt/json.c:
```diff
@@ -338,12 +354,67 @@ datum_to_json(Datum val, bool is_null, StringInfo result,
+	char *before_space;
+	char *before_utc;
+	char *end_str;
+	char *begin_space = strstr(outputstr, " ");
```
Please consider corner cases where `outputstr` or `begin_space` is null.

anujsarda3 committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV__PG_14_X' into jira-babel-3691

anujsarda3 committed:
Undo the change in json.c

- Undo the changes for the support for datatypes

TASK: BABEL-3691
Signed-off-by: Anuj Sarda<sardanuj@amazon.com>

anujsarda3 committed:
Undo the change in json.c

- Undo the changes for the support for datatypes
- Removed enum declaration from parser.h

TASK: BABEL-3691
Signed-off-by: Anuj Sarda<sardanuj@amazon.com>

anujsarda3 committed:
Merge branch 'jira-babel-3691' of https://github.com/amazon-aurora/postgresql_modified_for_babelfish into jira-babel-3691

anujsarda3 committed:
Nit edits in json.c

- Nit edits in json.c

TASK: BABEL-3691
Signed-off-by: Anuj Sarda<sardanuj@amazon.com>
